src/alire/alire-uri.ads

```diff
@@ -31,45 +31,126 @@ package Alire.URI with Preelaborate is
+      --  http/https, but not a recognized VCS
 
-      System,
-      --  system:package is used to denote a native package from the platform
+      SSH_Other,
+      --  ssh, but not a recognized VCS
 
       Unknown
       --  Anything else
      );
-   --  Protocols recognized by Alire
 
-   subtype VCS_Schemes is Schemes range Git .. SVN;
+   subtype Local_Other is URI_Kinds range Bare_Path .. File;
+
+   subtype Public_Git is URI_Kinds
+     range Public_Probably_Git .. Public_Definitely_Git;
+   --  Note that this includes the ambiguous case Public_Probably_Git
+
+   subtype Private_Git is URI_Kinds range SCP_Style_Git .. Other_Git;
+
+   subtype Non_URLs is URI_Kinds
+     with Static_Predicate =>
+       Non_URLs in External | System | SCP_Style_Git | Bare_Path | Unknown;
 
-   subtype File_Schemes is Schemes with
-     Static_Predicate => File_Schemes in None | File;
+   subtype VCS_URIs is URI_Kinds range Local_Git .. Private_SVN;
+   --  Note that this includes the ambiguous case Public_Probably_Git
 
-   function Scheme (This : URL) return Schemes;
-   --  Extract the Scheme part of a URL
+   subtype Local_VCS_URIs is URI_Kinds
+     with Static_Predicate =>
+       Local_VCS_URIs in Local_Git | Local_Hg | Local_SVN;
+
+   subtype Local_URIs is URI_Kinds
+     with Static_Predicate =>
+       Local_URIs in Local_VCS_URIs | Local_Other;
+
+   subtype Public_VCS_URIs is URI_Kinds
+     with Static_Predicate =>
+       Public_VCS_URIs in Public_Git | Public_Hg | Public_SVN;
+   --  Note that this includes the ambiguous case Public_Probably_Git
+
+   subtype Public_URIs is URI_Kinds
+     with Static_Predicate =>
+       Public_URIs in Public_VCS_URIs | Public_Other;
+
+   subtype Private_VCS_URIs is URI_Kinds
+     with Static_Predicate =>
+       Private_VCS_URIs in Private_Git | Private_Hg | Private_SVN;
+
+   subtype Private_URIs is URI_Kinds
+     with Static_Predicate =>
+       Private_URIs in Private_VCS_URIs | SSH_Other;
+
+   subtype Git_URIs is VCS_URIs range Local_Git .. Other_Git;
+   --  Note that this includes the ambiguous case Public_Probably_Git
+
+   subtype Hg_URIs is VCS_URIs range Local_Hg .. Private_Hg;
+
+   subtype SVN_URIs is VCS_URIs range Local_SVN .. Private_SVN;
+
+   function URI_Kind (This : String) return URI_Kinds;
+   --  Attempt to identify the nature of a resource from its URI.
+   --
+   --  Formats currently not recognized include:
+   --    user@host:/path/to/repo.git      [returns Unknown]
+   --    host.name:/path/to/repo.git      [returns Unknown]
+   --    git://host/path/to/repo.git      [returns Unknown]
+   --    ftp://host/path/to/repo.git      [returns Unknown]
+   --    svn://(something)                [returns Unknown]
+   --    https://user:pass@host/repo.git  [returns Public_Git]
+   --    https://user@host/repo.git       [returns Public_Git]
+   --    https://user:pass@host/path      [returns Public_Other]
+   --    https://user@host/path           [returns Public_Other]
```

Comment:
As a general rule, I have tried to preserve existing behaviour, except where said behaviour was inconsistent. There was previously no distinction between `https` URLs with or without a userinfo component, so I have kept that the same.

If you want, I could change `Alire.URI_Kind` to treat `http[s]` URLs with a non-empty userinfo component as private, but I haven't really looked into whether this would break anything.

Reply:
I see. Well, as long as you have this fresh in your mind, I think it's a good opportunity to plug some holes. In particular, the ones returning `Public_*` could return `Private_*` and I don't expect it should affect anything. If you find unexpected troubles we could decide on leaving it as-is.